### PR TITLE
Add docs for UdpBatchSender

### PR DESCRIPTION
**Pull request by tphnicholas**
What should the `class` comment contain and what stays in the `description` section?

Are the comments descriptive enough?

### CODE REVIEW

**Comment by briantomko** [comment]
@tphnicholas These comments in the `UdpBatchSender `look great! I suppose the description and class comments could be identical, unless there's something specific that could be added to the class comment that makes sense.

**Comment by tphnicholas** [comment]
@briantomko I see, the example in `TokenRateLimiter` was more of a quick class usage showcase, was unsure if that's what you were going for. Would you like to sort this class comment out while reviewing the rest?

What do you think about documenting `@post`conditions at least for the public interface? It worked well in this case, not sure how it'll hold for more complex/involved classes, would be open to giving it a try.

**Comment by briantomko** [comment]
@tphnicholas I would say class comments should be necessary if there are multiple classes in a file, but not otherwise.  Also, how were you wanting to do the pull requests (one file per pull or several/all files per pull)?  I can review the UdpBatchSender first thing in the morning.

**Comment by tphnicholas** [comment]
@briantomko I was thinking grouping by sub-components for the main network components of DTN should make sense and keep the history kind of clean.

For example starting with 1 PR for the UDP layer of ltp-over-udp which would include the udp engine, ltp-over-udp induct/outduct, the delay simulation utility and this file, thoughts?

**Comment by briantomko** [comment]
> @briantomko I was thinking grouping by sub-components for the main network components of DTN should make sense and keep the history kind of clean.
> 
> For example starting with 1 PR for the UDP layer of ltp-over-udp which would include the udp engine, ltp-over-udp induct/outduct, the delay simulation utility and this file, thoughts?

@tphnicholas That sounds good.  Just let me know when you feel the PR is ready for merge so I don't accidentally merge it too soon.

File: common/util/include/UdpBatchSender.h [hunk 1]
```diff
@@ -32,50 +32,151 @@
...
+     * @return The current UDP endpoint.
+     */
     HDTN_UTIL_EXPORT boost::asio::ip::udp::endpoint GetCurrentUdpEndpoint() const;
 
+    /** Initiate a packet batch send operation (thread-safe).
+     *
+     * Initiates an asynchronous request to UdpBatchSender::PerformSendPacketsOperation().
+     * @param constBufferVecs The packets to send.
+     * @param underlyingDataToDeleteOnSentCallbackVec (internal) Vector of LtpClientServiceDataToSend::underlyingDataToDeleteOnSentCallback.
+     * @param underlyingCsDataToDeleteOnSentCallbackVec (internal) Vector of LtpClientServiceDataToSend::underlyingCsDataToDeleteOnSentCallback.
```
Comment (briantomko) [inline common/util/include/UdpBatchSender.h:82]: Should add a post condition that all three parameter vectors are moved/stolen, so after the function call, the vectors are left in an unspecified, but valid state (i.e. `vec.empty()` will return `true`)

File: common/util/include/UdpBatchSender.h [hunk 2]
```diff
@@ -32,50 +32,151 @@
...
+     *
+     * @return The current UDP endpoint.
+     */
     HDTN_UTIL_EXPORT boost::asio::ip::udp::endpoint GetCurrentUdpEndpoint() const;
 
+    /** Initiate a packet batch send operation (thread-safe).
+     *
+     * Initiates an asynchronous request to UdpBatchSender::PerformSendPacketsOperation().
+     * @param constBufferVecs The packets to send.
+     * @param underlyingDataToDeleteOnSentCallbackVec (internal) Vector of LtpClientServiceDataToSend::underlyingDataToDeleteOnSentCallback.
```
Comment (briantomko) [inline common/util/include/UdpBatchSender.h:81]: This is the vector of raw bytes (referenced by the constBufferVecs) that must be persistent until after the OnSentPacketsCallback_t is called, and is generally used for smaller data pieces of the udp packet such as ltp headers.

File: common/util/include/UdpBatchSender.h (hunk 1 above)
Comment (briantomko) [inline common/util/include/UdpBatchSender.h:82]: This is the vector of raw bytes from the LtpClientServiceDataToSend (referenced by the constBufferVecs) that must be persistent until after the OnSentPacketsCallback_t is called, and its general use is documented here: https://github.com/nasa/HDTN/blob/master/common/ltp/src/LtpSessionSender.cpp#L268

File: common/util/include/UdpBatchSender.h [hunk 3]
```diff
@@ -32,50 +32,151 @@
...
     HDTN_UTIL_NO_EXPORT bool SetEndpointAndReconnect(const std::string& remoteHostname, const uint16_t remotePort);
+    
+    /** Perform a packet batch send operation.
+     *
+     * Performs a single synchronous batch send operation.
+     * After it finishes (successfully or otherwise), the callback stored in m_onSentPacketsCallback (if any) is invoked.
+     * Arguments underlyingDataToDeleteOnSentCallbackVec and underlyingCsDataToDeleteOnSentCallbackVec are resources
+     * managed by LtpClientServiceDataToSend, we should simply forward them as-is.
+     * @param constBufferVecs The packets to send.
+     * @param underlyingDataToDeleteOnSentCallbackVec (internal) Vector of LtpClientServiceDataToSend::underlyingDataToDeleteOnSentCallback.
```
Comment (briantomko) [inline common/util/include/UdpBatchSender.h:132]: Correct this to above comment

**Review by briantomko — CHANGES_REQUESTED** [review]

**Comment by tphnicholas** [comment]
@briantomko You could close this one out and I incorporate this file + suggested changes over to a new PR once the rest of the files are ready for review to keep everything bundled together.

**Comment by briantomko** [comment]
> @briantomko You could close this one out and I incorporate this file + suggested changes over to a new PR once the rest of the files are ready for review to keep everything bundled together.

Yes I'll close this out, and the review I did here can go to the new PR/issue.

**Pull request closed without merging**